### PR TITLE
More thunk bashing

### PR DESCRIPTION
**Pull request by mrBliss**

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Types.hs [hunk 1]
```diff
@@ -207,7 +208,14 @@ data ChainDbEnv m blk = CDB
...
   , cdbEpochInfo      :: !(EpochInfo m)
-  } deriving (Generic, NoUnexpectedThunks)
+  } deriving (Generic)
+
+-- | We include @blk@ in 'showTypeOf' because it helps resolving type families
+-- (but avoid including @m@ because we cannot impose @Typeable m@ as a
+-- constraint and still have it work with the simulator)
+instance (IOLike m, ProtocolLedgerView blk)
+      => NoUnexpectedThunks (ChainDbEnv m blk) where
+    showTypeOf _ = "ChainDbEnv m " ++ show (typeRep (Proxy @blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Types.hs]: Yeah, I meant to ask you about when to override `showTypeOf`, I ignored it so far.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs [hunk 2]
```diff
@@ -148,10 +152,22 @@ isEmptyMVar StrictMVar { tmvar } = atomically $ Lazy.isEmptyTMVar tmvar
 
 updateMVar :: MonadSTM m => StrictMVar m a -> (a -> (a, b)) -> m b
 updateMVar StrictMVar { tmvar, tvar, invariant } f = do
-    (a', b) <- atomically $ do
+    (!a', b) <- atomically $ do
         a <- Lazy.takeTMVar tmvar
         let (!a', b) = f a
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs]: So we should put a bang on the outside of this tuple too?

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 3]
```diff
@@ -820,7 +824,13 @@ data Context m = IOLike m => Context {
     , contextThreadId  :: !(ThreadId m)
     }
 
-deriving via UseIsNormalFormNamed "Context" (Context m) instance NoUnexpectedThunks (Context m)
+instance NoUnexpectedThunks (Context m) where
+  showTypeOf _ = "Context"
+  whnfNoUnexpectedThunks ctxt (Context cs _tid) = allNoUnexpectedThunks
+    [ noUnexpectedThunks ctxt cs
+      -- TODO tid?
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: TODO

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Iterator.hs [hunk 4]
```diff
@@ -452,8 +456,8 @@ data IteratorState m blk
     --
     -- Invariant: the iterator is not exhausted.
   | InVolDB
-      (StreamFrom blk)
-      (NonEmpty (HeaderHash blk))
+      !(StreamFrom blk)
+      !(NonEmpty (HeaderHash blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Iterator.hs]: I need to investigate why this isn't giving us thunks (see other constructors of this type).

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs]: ...instead of the current change.

**Comment by edsko** [comment]
Obsoleted by #1114.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs (hunk 2 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs]: I don't think so. It's a pattern match in a lambda, so not a lazy pattern.

**Pull request closed without merging**